src/aspire/utils/blk_diag_matrix.py

```diff
@@ -0,0 +1,700 @@
+         self + other.
+        """
+        assert self._is_scalar_type(scalar)
+
+        if not inplace:
+            C = BlkDiagMatrix.from_blk_diag(self.data)
+        else:
+            C = self
+
+        for i in range(self.nblocks):
+            C[i] -= scalar
+
+        return C
+
+    def matmul(self, other):
+        """
+        Compute the Matrix multiplication of two BlkDiagMatrix matrices
+
+        :param other: The rhs BlkDiagMatrix matrix
+
+        :return: A BlkDiagMatrix of self @ other.
+        """
+        if not isinstance(other, BlkDiagMatrix):
+            raise RuntimeError(
+                "Attempt BlkDiagMatrix Matrix multiplication "
+                "(matmul,@) of non BlkDiagMatrix {}, try (*,mul)".format(
+                    repr(other)))
+
+        self.check_compatible(other)
+
+        C = BlkDiagMatrix(self.nblocks, dtype=self.dtype)
+
+        for i in range(self.nblocks):
+            C[i] = self[i] @ other[i]
+        return C
+
+    def __matmul__(self, other):
+        """ Operator overload for matrix matrix multiply of BlkDiagMatrixes """
+        return self.matmul(other)
+
+    def __imatmul__(self, other):
+        """
+        Operator overload for in place matrix matrix multiply
+        of BlkDiagMatrixes.
+        """
+        self.check_compatible(other)
+        for i in range(self.nblocks):
+            self[i] @= other[i]
+
+        return self
+
+    def mul(self, val):
+        """
+        Compute the numeric (elementwise) multiplication of a BlkDiagMatrix
+         matrix and a scalar.
+
+        :param other: The rhs BlkDiagMatrix matrix
+
+        :return: A BlkDiagMatrix of self * other.
+        """
+
+        if isinstance(val, BlkDiagMatrix):
+            raise RuntimeError("Attempt numeric multiplication (*,mul) of two "
+                               "BlkDiagMatrixs, try (matmul,@)")
+        elif not self._is_scalar_type(val):
+            raise RuntimeError("Attempt numeric multiplication (*,mul) of a "
+                               "BlkDiagMatrix and {}".format(
+                                   type(val)))
+
+        C = BlkDiagMatrix(self.nblocks, dtype=self.dtype)
+
+        for i in range(self.nblocks):
+            C[i] = self[i] * val
+        return C
+
+    def __mul__(self, val):
+        """ Operator overload for BlkDiagMatrix scalar multiply"""
+        return self.mul(val)
+
+    def __imul__(self, val):
+        """ Operator overload for in place BlkDiagMatrix scalar multiply"""
+        if isinstance(val, BlkDiagMatrix):
+            raise RuntimeError(
+                "Attempt numeric multiplication (*,mul) of two BlkDiagMatrixs,"
+                " try (matmul,@)")
+        for i in range(self.nblocks):
+            self[i] *= val
+
+        return self
+
+    def __rmul__(self, other):
+        """ Convenience function, elementwise rmul commutes to mul """
+        return self.mul(other)
+
+    def __neg__(self):
```

Comment:
No its just one flavor of writing.  Another being numpy, a+b and np.add(a,b).

I often will write it this way if I think it would be optimized later, in which case __neg__ might be the only thing left in pure python...

Simple to change, just let me know what you want to see (sounds like only __op__)...

Reply:
I misunderstood what you meant earlier. Added neg, pow, abs methods to be complete. Resolving.

Reply:
Yes that's what I meant. Sorry for being unclear.